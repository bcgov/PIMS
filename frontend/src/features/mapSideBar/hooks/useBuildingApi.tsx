import { request, success, error } from 'actions/genericActions';
import { IBuilding } from 'actions/parcelsActions';
import { showLoading, hideLoading } from 'react-redux-loading-bar';
import CustomAxios, { LifecycleToasts } from 'customAxios';
import { ENVIRONMENT } from 'constants/environment';
import * as actionTypes from 'constants/actionTypes';
import * as pimsToasts from 'constants/toasts';
import * as API from 'constants/API';
import { fetchBuildingDetail } from 'actionCreators/parcelsActionCreator';

const buildingCreatingToasts: LifecycleToasts = {
  loadingToast: pimsToasts.building.BUILDING_CREATING,
  successToast: pimsToasts.building.BUILDING_CREATED,
  errorToast: pimsToasts.building.BUILDING_CREATING_ERROR,
};

const buildingUpdatingToasts: LifecycleToasts = {
  loadingToast: pimsToasts.building.BUILDING_UPDATING,
  successToast: pimsToasts.building.BUILDING_UPDATED,
  errorToast: pimsToasts.building.BUILDING_UPDATING_ERROR,
};

export const useBuildingApi = () => {
  /**
   * Create the passed building, creating or updating all attached parcels as needed. Return a promise
   */
  const createBuilding = (building: IBuilding) => async (dispatch: Function) => {
    dispatch(request(actionTypes.ADD_BUILDING));
    dispatch(showLoading());
    try {
      const { data, status } = await CustomAxios({ lifecycleToasts: buildingCreatingToasts }).post(
        ENVIRONMENT.apiUrl + API.BUILDING_ROOT,
        building,
      );
      dispatch(success(actionTypes.ADD_BUILDING, status));
      dispatch(hideLoading());
      return data;
    } catch (axiosError) {
      dispatch(error(actionTypes.ADD_BUILDING, axiosError?.response?.status, axiosError));
      dispatch(hideLoading());
      throw Error(axiosError.response?.data.details);
    }
  };

  /**
   * update the passed building (requires an id and rowversions). Also create/update all associated parcels as needed.
   * @param building
   */
  const updateBuilding = (building: IBuilding) => async (dispatch: Function) => {
    dispatch(request(actionTypes.UPDATE_BUILDING));
    dispatch(showLoading());
    try {
      const { data, status } = await CustomAxios({ lifecycleToasts: buildingUpdatingToasts }).put(
        ENVIRONMENT.apiUrl + API.BUILDING_ROOT + `/${building.id}`,
        building,
      );
      dispatch(success(actionTypes.UPDATE_BUILDING, status));
      dispatch(fetchBuildingDetail({ id: +building.id }));
      dispatch(hideLoading());
      return data;
    } catch (axiosError) {
      dispatch(error(actionTypes.UPDATE_BUILDING, axiosError?.response?.status, axiosError));
      dispatch(hideLoading());
      throw Error(axiosError.response?.data.details);
    }
  };

  return { createBuilding, updateBuilding };
};
