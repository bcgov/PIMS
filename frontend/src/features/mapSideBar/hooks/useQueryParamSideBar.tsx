import { useLocation, useHistory } from 'react-router-dom';
import { useState, useCallback, useMemo } from 'react';
import queryString from 'query-string';
import useDeepCompareEffect from 'hooks/useDeepCompareEffect';
import * as H from 'history';
import dequal from 'dequal';

export type SidebarSize = 'narrow' | 'wide' | undefined;

export enum SidebarContextType {
  ADD_PROPERTY_TYPE_SELECTOR = 'addPropertyTypeSelector',
  ADD_BUILDING = 'addBuilding',
  ADD_BARE_LAND = 'addBareLand',
  ADD_ASSOCIATED_LAND = 'addAssociatedLand',
  ADD_SUBDIVISION_LAND = 'addSubdivisionLand',
  VIEW_BUILDING = 'viewBuilding',
  VIEW_BARE_LAND = 'viewBareLand',
  VIEW_DEVELOPED_LAND = 'viewDevelopedLand',
  VIEW_SUBDIVISION_LAND = 'viewSubdivisionLand',
  UPDATE_BUILDING = 'updateBuilding',
  UPDATE_BARE_LAND = 'updateBareLand',
  UPDATE_DEVELOPED_LAND = 'updateDevelopedLand',
  UPDATE_SUBDIVISION_LAND = 'updateSubdivisionLand',
  LOADING = 'loading',
}

interface IMapSideBar {
  showSideBar: boolean;
  setShowSideBar: (
    show: boolean,
    contextName?: SidebarContextType,
    size?: SidebarSize,
    resetParcelId?: boolean,
  ) => void;
  addBuilding: () => void;
  addBareLand: () => void;
  addAssociatedLand: () => void;
  addSubdivision: () => void;
  addContext: (context: SidebarContextType) => void;
  setDisabled: (disabled: boolean) => void;
  parcelId?: number;
  buildingId?: number;
  associatedParcelId?: number;
  disabled?: boolean;
  loadDraft?: boolean;
  newParcel?: boolean;
  size?: SidebarSize;
  context: SidebarContextType;
  handleLocationChange:
    | string
    | ((location: H.Location, action: 'PUSH' | 'POP' | 'REPLACE') => string | boolean);
}

/** control the state of the side bar via query params. */
export const useQueryParamSideBar = (formikRef?: any): IMapSideBar => {
  const [showSideBar, setShowSideBar] = useState(false);
  const [contextName, setContextName] = useState<SidebarContextType>(
    SidebarContextType.ADD_PROPERTY_TYPE_SELECTOR,
  );
  const [sideBarSize, setSideBarSize] = useState<SidebarSize>(undefined);
  const [parcelId, setParcelId] = useState<number | undefined>(undefined);
  const [buildingId, setBuildingId] = useState<number | undefined>(undefined);
  const [associatedParcelId, setAssociatedParcelId] = useState<number | undefined>(undefined);
  const location = useLocation();
  const history = useHistory();

  const searchParams = useMemo(() => queryString.parse(location.search), [location.search]);
  useDeepCompareEffect(() => {
    setShowSideBar(searchParams.sidebar === 'true');
    setParcelId(searchParams.parcelId ? +searchParams.parcelId || undefined : undefined);
    setBuildingId(
      searchParams.buildingId !== undefined && searchParams.buildingId !== null
        ? +searchParams.buildingId
        : undefined,
    );
    setAssociatedParcelId(
      searchParams.associatedParcelId ? +searchParams.associatedParcelId || undefined : undefined,
    );
    setSideBarSize(searchParams.sidebarSize as SidebarSize);
    setContextName(searchParams.sidebarContext as SidebarContextType);
    if (searchParams?.new === 'true') {
      const queryParams: any = { ...searchParams, new: false };
      queryParams.parcelId = undefined;
      queryParams.buildingId = undefined;
      queryParams.sidebarContext = SidebarContextType.ADD_PROPERTY_TYPE_SELECTOR;
      history.replace({ pathname: '/mapview', search: queryString.stringify(queryParams) });
    } else if (
      searchParams.sidebar === 'false' &&
      (searchParams.parcelId || searchParams.buildingId || searchParams.associatedParcelId)
    ) {
      searchParams.parcelId = undefined;
      searchParams.buildingId = undefined;
      searchParams.associatedParcelId = undefined;
      history.replace({
        pathname: '/mapview',
        search: queryString.stringify(searchParams),
      });
    }
  }, [searchParams]);

  const setShow = useCallback(
    (show: boolean, contextName?: SidebarContextType, size?: SidebarSize, resetIds?: boolean) => {
      if (show && !contextName) {
        throw new Error('"contextName" is required when "show" is true');
      }

      const search = {
        ...(searchParams as any),
        sidebar: show,
        sidebarSize: show ? size : undefined,
        sidebarContext: show ? contextName : undefined,
        parcelId: resetIds ? undefined : searchParams.parcelId,
        buildingId: resetIds ? undefined : searchParams.buildingId,
      };
      history.push({ search: queryString.stringify(search) });
    },
    [history, searchParams],
  );

  const addBuilding = () => {
    setShow(true, SidebarContextType.ADD_BUILDING, 'wide');
  };

  const addBareLand = () => {
    setShow(true, SidebarContextType.ADD_BARE_LAND, 'wide');
  };

  const addAssociatedLand = () => {
    setShow(true, SidebarContextType.ADD_ASSOCIATED_LAND, 'wide');
  };

  const addSubdivision = () => {
    setShow(true, SidebarContextType.ADD_SUBDIVISION_LAND, 'wide');
  };

  const addContext = useCallback(
    (context: SidebarContextType) => {
      setShow(showSideBar, context, 'wide');
    },
    [setShow, showSideBar],
  );

  const handleLocationChange = (location: H.Location, action: 'PUSH' | 'POP' | 'REPLACE') => {
    const parsedChangedLocation = queryString.parse(location.search);
    return (searchParams.sidebarContext !== parsedChangedLocation.sidebarContext ||
      searchParams.parcelId !== parsedChangedLocation.parcelId ||
      searchParams.buildingId !== parsedChangedLocation.buildingId ||
      searchParams.sideBar !== parsedChangedLocation.sideBar) &&
      !dequal(formikRef?.current?.initialValues?.data, formikRef?.current?.values?.data)
      ? 'Are you sure you want to leave this page? You have unsaved changes.'
      : true;
  };

  return {
    showSideBar,
    context: contextName,
    setShowSideBar: setShow,
    size: sideBarSize,
    parcelId,
    buildingId,
    associatedParcelId,
    disabled: searchParams?.disabled === 'true',
    loadDraft: searchParams?.loadDraft === 'true',
    newParcel: searchParams?.new === 'true',
    addBuilding,
    addBareLand,
    addAssociatedLand,
    addSubdivision,
    addContext,
    setDisabled: disabled => {
      const queryParams = {
        ...queryString.parse(location.search),
        loadDraft: true,
        disabled: disabled,
      };
      const pathName = '/mapview';
      history.replace({ pathname: pathName, search: queryString.stringify(queryParams) });
    },
    handleLocationChange,
  };
};

export default useQueryParamSideBar;
