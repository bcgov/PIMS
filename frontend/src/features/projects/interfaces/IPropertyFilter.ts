/**
 * IPropertyFilter interface, provides a model for querying the API for properties.
 */
export interface IPropertyFilter {
  page: number;
  quantity: number;
  pid?: string;
  address?: string;
  administrativeArea?: string;
  projectNumber?: string;
  ignorePropertiesInProjects?: boolean;
  inSurplusPropertyProgram?: boolean;
  classificationId?: number;
  agencies?: number | number[];
  minLandArea?: number;
  maxLandArea?: number;
  minLotArea?: number;
  maxLotArea?: number;
  all?: boolean;
}
