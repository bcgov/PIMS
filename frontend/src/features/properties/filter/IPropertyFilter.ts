/**
 * Property filter options used by Formik.
 */
export interface IPropertyFilter {
  /** Select one of the search options [address, pid, projectNumber, location]. */
  searchBy: string;
  /** The page number. */
  page?: string;
  /** The quantity to return in a single request for paging. */
  quantity?: string;
  /** The unique PID of the parcel. */
  pid: string;
  /** The address of the property. */
  address: string;
  /** The location of the property. */
  administrativeArea: string;
  /** The project number the property is part of. */
  projectNumber: string;
  /** comma-separated list of agencies to filter by */
  agencies: string;
  /** The classification of the property. */
  classificationId: string;
  /** The minimum lot size of the property. */
  minLotSize: string;
  /** The maxium lot size of the property. */
  maxLotSize: string;
  /** Whether the property is in SPP. */
  inSurplusPropertyProgram?: boolean;
  /** Whether the property is in ERP. */
  inEnhancedReferralProcess?: boolean;
  /** Select on of the property types [Land, Building]. */
  propertyType?: string;
  /** The name of desired target */
  name?: string;
  /** The building construction type id */
  constructionTypeId?: string;
  /** The building predominant use id */
  predominateUseId?: string;
  /** The building number of floors */
  floorCount?: string;
  /** Flag for whether to include buildings */
  bareLandOnly?: boolean;
  /** filter for building rentable area */
  rentableArea: string;
  /** The maximum Assesses Value for a property */
  maxAssessedValue?: string;
  /** The maximum Net book Value for a property */
  maxNetBookValue?: '';
  /** The maximum Market Value for a property */
  maxMarketValue?: '';
  /** Whether to return properties owned by other agencies. */
  includeAllProperties?: boolean;
  /** when the surplus properties filter is open */
  surplusFilter?: boolean;
}
