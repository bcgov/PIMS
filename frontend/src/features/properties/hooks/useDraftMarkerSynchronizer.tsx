import * as React from 'react';
import { PropertyTypes, storeDraftParcelsAction } from 'actions/parcelsActions';
import useDeepCompareEffect from 'hooks/useDeepCompareEffect';
import debounce from 'lodash/debounce';
import { useFormikContext, getIn } from 'formik';
import { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import _ from 'lodash';
import { RootState } from 'reducers/rootReducer';
import { PointFeature } from 'components/maps/types';

/**
 * Get a list of draft markers from the current form values.
 * As long as a parcel/building has both a lat and a lng it will be returned by this method.
 * @param values the current form values to extract lat/lngs from.
 * @param initialValues the original form values, used to exclude unchanged lat/lngs
 * @param nameSpace path within above objects to extract lat/lngs
 */
const getDraftMarkers = (values: any, initialValues: any, nameSpace: string) => {
  values = getIn(values, nameSpace);
  initialValues = getIn(initialValues, nameSpace);
  if (
    values?.latitude === '' ||
    values?.longitude === '' ||
    (values?.latitude === initialValues?.latitude && values?.longitude === initialValues?.longitude)
  ) {
    return [];
  }
  return [
    {
      type: 'Feature',
      geometry: {
        type: 'Point',
        coordinates: [+values.longitude, +values.latitude],
      },
      properties: {
        id: 0,
        name: values.name?.length ? values.name : 'New Parcel',
        propertyTypeId:
          values.parcelId !== undefined ? PropertyTypes.DRAFT_BUILDING : PropertyTypes.DRAFT_PARCEL,
      },
    },
  ];
};

/**
 * A hook that automatically syncs any updates to the lat/lngs of the parcel form with the map.
 * @param param0 The currently displayed list of properties on the map.
 */
const useDraftMarkerSynchronizer = (nameSpace: string) => {
  const { values, initialValues } = useFormikContext();
  const properties = useSelector<RootState, PointFeature[]>(state => [
    ...state.parcel.draftParcels,
  ]);
  const dispatch = useDispatch();
  const nonDraftProperties = React.useMemo(
    () =>
      properties.filter(
        (property: PointFeature) =>
          property.properties.propertyTypeId !== undefined &&
          [PropertyTypes.BUILDING, PropertyTypes.PARCEL].includes(
            property.properties.propertyTypeId,
          ),
      ),
    [properties],
  );

  React.useEffect(() => {
    return () => {
      dispatch(storeDraftParcelsAction([]));
    };
  }, [dispatch]);

  /**
   * Synchronize the markers that have been updated in the parcel form with the map, adding all new markers as drafts.
   * @param values the current form values
   * @param initialValues the initial form values
   * @param dbProperties the currently displayed list of (DB) map properties.
   * @param nameSpace the path to extract lat/lng values from
   */
  const synchronizeMarkers = (
    values: any,
    initialValues: any,
    dbProperties: PointFeature[],
    nameSpace: string,
  ) => {
    const draftMarkers = getDraftMarkers(values, initialValues, nameSpace);
    if (draftMarkers.length) {
      const newDraftMarkers = _.filter(
        draftMarkers,
        (draftMarker: PointFeature) =>
          _.find(
            dbProperties,
            dbProperty =>
              dbProperty.geometry.coordinates[0] === draftMarker.geometry.coordinates[0] &&
              dbProperty.geometry.coordinates[1] === draftMarker.geometry.coordinates[1],
          ) === undefined,
      );
      dispatch(storeDraftParcelsAction(newDraftMarkers as PointFeature[]));
    } else {
      dispatch(storeDraftParcelsAction([]));
    }
  };

  const synchronize = useCallback(
    debounce((values: any, initialValues: any, properties: PointFeature[], nameSpace: string) => {
      synchronizeMarkers(values, initialValues, properties, nameSpace);
    }, 400),
    [],
  );

  useDeepCompareEffect(() => {
    synchronize(values, initialValues, nonDraftProperties, nameSpace);
  }, [values, initialValues, nonDraftProperties, synchronize, nameSpace]);

  return;
};

export default useDraftMarkerSynchronizer;
