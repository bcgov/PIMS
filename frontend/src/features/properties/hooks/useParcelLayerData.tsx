import { FormikValues, getIn, setIn } from 'formik';
import { isMouseEventRecent, squareMetersToHectares } from 'utils';
import { AMINISTRATIVE_AREA_CODE_SET_NAME } from 'constants/API';
import { useState } from 'react';
import { IParcelLayerData } from 'reducers/parcelLayerDataSlice';
import { ILookupCode } from 'actions/lookupActions';
import { pidFormatter } from '../components/forms/subforms/PidPinForm';
import _ from 'lodash';
import { useSelector } from 'react-redux';
import { RootState } from 'reducers/rootReducer';
import useCodeLookups from 'hooks/useLookupCodes';
import { toast } from 'react-toastify';
import useDeepCompareEffect from 'hooks/useDeepCompareEffect';
import { getInitialValues } from 'features/mapSideBar/SidebarContents/LandForm';

interface IUseParcelLayerDataProps {
  formikRef: React.MutableRefObject<FormikValues | undefined>;
  agencyId: number;
  parcelId?: number | '';
  nameSpace?: string;
}

/**
 * Only autofill data if:
 * 1) layerdata and formikref are bot non-null/undefined
 * 2) either the layer pid or pin has a non-zero value
 * 3) the pid or pin are not equal to the initial values (indicating that the form has already been pre-populated)
 * @param layerData
 * @param formikRef
 * @param nameSpace
 */
const isFormInStateToSetLayerData = (
  layerData: IParcelLayerData | null,
  formikRef: React.MutableRefObject<FormikValues | undefined>,
  nameSpace: string,
) => {
  const formPid = getIn(formikRef?.current?.initialValues, `${nameSpace}.pid`)?.replace(/-/g, '');
  const formPin = getIn(formikRef?.current?.initialValues, `${nameSpace}.pin`);
  const pid = +layerData?.data.PID || +layerData?.data?.PID_NUMBER;
  return (
    !!formikRef.current &&
    !!layerData?.data &&
    ((pid > 0 && +formPid !== pid) || (+layerData.data.PIN > 0 && +formPin !== +layerData.data.PIN))
  );
};

/**
 * Set the fields in the parcel form using the passed formik reference.
 * @param layerData The parcel layer data to use to set the data
 * @param formikRef The formik reference to the parcel detail form.
 * @param administrativeAreas the full list of administrative areas, used to find an admin area matching the code set.
 */
const setParcelFieldsFromLayerData = (
  layerData: IParcelLayerData | null,
  formikRef: React.MutableRefObject<FormikValues | undefined>,
  administrativeAreas: ILookupCode[],
  nameSpace: string,
  agencyId: number,
) => {
  if (isFormInStateToSetLayerData(layerData, formikRef, nameSpace)) {
    toast.dark('Autofilling form utilizing BC Geographic Warehouse data.', { autoClose: 7000 });
    const { values, setValues } = formikRef.current!;
    let newValues = { ...values };
    if (nameSpace) {
      newValues = setIn(newValues, nameSpace, {
        ...getInitialValues(),
        ...getIn(values, nameSpace),
      });
    }

    const layerParcelData = layerData!.data;

    //These fields are generic and used in both buildings and parcels
    newValues = setIn(
      newValues,
      `${nameSpace}.landArea`,
      squareMetersToHectares(+layerParcelData.FEATURE_AREA_SQM),
    );
    newValues = setIn(newValues, `${nameSpace}.latitude`, layerParcelData.CENTER.lat);
    newValues = setIn(newValues, `${nameSpace}.longitude`, layerParcelData.CENTER.lng);
    const administrativeArea = getAdminAreaFromLayerData(
      administrativeAreas,
      layerParcelData.MUNICIPALITY,
    );
    if (administrativeArea) {
      newValues = setIn(
        newValues,
        `${nameSpace}.address.administrativeArea`,
        administrativeArea.name,
      );
    }

    //parcel only fields.
    if (getIn(values, `${nameSpace}.pid`) !== undefined) {
      const pid = layerData?.data?.PID || layerData?.data?.PID_NUMBER?.toString();
      newValues = setIn(newValues, `${nameSpace}.pid`, !!pid ? pidFormatter(pid) : '');
      newValues = setIn(newValues, `${nameSpace}.pin`, layerParcelData.PIN || '');

      newValues = setIn(newValues, `${nameSpace}.agencyId`, agencyId);
      const searchAddress = getIn(values, `${nameSpace}.searchAddress`);
      if (searchAddress && getIn(newValues, `${nameSpace}.address.line1`) === '') {
        newValues = setIn(newValues, `${nameSpace}.address.line1`, searchAddress);
      }
      newValues = setIn(
        newValues,
        `${nameSpace}.searchPin`,
        getIn(values, `${nameSpace}.searchPin`),
      );
      newValues = setIn(
        newValues,
        `${nameSpace}.searchPid`,
        getIn(values, `${nameSpace}.searchPid`),
      );
      newValues = setIn(newValues, `${nameSpace}.searchAddress`, searchAddress);
    }
    setValues({ ...values, ...newValues });
  }
};

/**
 * Using the administrative areas code set, find the matching municipality returned by the parcel layer, if present.
 * @param administrativeAreas the full list from the administrative areas code set.
 * @param layerMunicipality the municipality returned by the layer.
 */
const getAdminAreaFromLayerData = (
  administrativeAreas: ILookupCode[],
  layerMunicipality: string,
) => {
  let administrativeArea = _.find(administrativeAreas, { name: layerMunicipality });
  if (administrativeArea) {
    return administrativeArea;
  }
  if (!!layerMunicipality?.length) {
    const splitLayerMunicipality = layerMunicipality.split(',');
    if (splitLayerMunicipality.length === 2) {
      const formattedLayerMunicipality = `${splitLayerMunicipality[1].trim()} ${splitLayerMunicipality[0].trim()}`;
      return _.find(administrativeAreas, { name: formattedLayerMunicipality });
    }
  }
};

/**
 * hook providing methods to update the parcel detail form using parcel layer data.
 */
const useParcelLayerData = ({
  formikRef,
  parcelId,
  nameSpace,
  agencyId,
}: IUseParcelLayerDataProps) => {
  const parcelLayerData = useSelector<RootState, IParcelLayerData | null>(
    state => state.parcelLayerData?.parcelLayerData,
  );
  const { getByType } = useCodeLookups();
  const [showOverwriteDialog, setShowOverwriteDialog] = useState(false);

  useDeepCompareEffect(() => {
    if (!!formikRef?.current && isMouseEventRecent(parcelLayerData?.e) && !!parcelLayerData?.data) {
      if (!parcelId) {
        setParcelFieldsFromLayerData(
          parcelLayerData,
          formikRef,
          getByType(AMINISTRATIVE_AREA_CODE_SET_NAME),
          nameSpace ?? '',
          agencyId,
        );
      } else {
        setShowOverwriteDialog(true);
      }
    }
  }, [formikRef, getByType, parcelId, parcelLayerData]);

  return {
    setShowOverwriteDialog,
    showOverwriteDialog,
    setParcelFieldsFromLayerData: () =>
      setParcelFieldsFromLayerData(
        parcelLayerData,
        formikRef,
        getByType(AMINISTRATIVE_AREA_CODE_SET_NAME),
        nameSpace ?? '',
        agencyId,
      ),
  };
};

export default useParcelLayerData;
