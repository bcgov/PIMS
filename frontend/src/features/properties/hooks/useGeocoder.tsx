import * as API from 'constants/API';
import { IGeocoderResponse, useApi } from 'hooks/useApi';
import { FormikValues, setIn, getIn } from 'formik';
import { useState } from 'react';
import useCodeLookups from 'hooks/useLookupCodes';
import {
  useLayerQuery,
  PARCELS_LAYER_URL,
  handleParcelDataLayerResponse,
} from 'components/maps/leaflet/LayerPopup';
import { LatLng } from 'leaflet';
import { useDispatch } from 'react-redux';
import { fetchParcelsDetail } from 'actionCreators/parcelsActionCreator';
import { toast } from 'react-toastify';
import _ from 'lodash';

interface IUseGeocoderProps {
  formikRef: React.MutableRefObject<FormikValues | undefined>;
}

export interface IPidSelection {
  /**
   * showPopup for when the geocoder pid does not match the current pid.
   */
  showPopup: boolean;
  /**
   * The pid returned by the geocoder for the current civil address.
   */
  geoPID: string;
}

/**
 * useGeocoder hook searches the geocoder api service whenever the civil address changes.
 * The parcel data layer may be searched based on the response latitude/longitude.
 * @param param0
 */
const useGeocoder = ({ formikRef }: IUseGeocoderProps) => {
  const { lookupCodes } = useCodeLookups();
  const parcelsService = useLayerQuery(PARCELS_LAYER_URL);
  const [pidSelection, setPidSelection] = useState<IPidSelection>({ showPopup: false, geoPID: '' });
  const api = useApi();
  const dispatch = useDispatch();

  const updateForm = (newValues: any, nameSpace?: string) => {
    if (!!formikRef?.current) {
      // update form with values returned from geocoder
      newValues.agencyId = formikRef.current.values.data.agencyId;
      if (!!nameSpace) {
        const { resetForm, values } = formikRef.current;
        resetForm({ values: setIn(values, nameSpace, newValues) });
      } else {
        formikRef.current.setValues(newValues);
      }
    }
  };

  const fetchPimsParcelByPid = (newValues: any, parcelPid: string, nameSpace?: string) => {
    updateForm(newValues, nameSpace);
    fetchParcelsDetail({ pid: parcelPid } as any)(dispatch).then(resp => {
      const matchingParcel: any = resp?.data?.length ? _.first(resp?.data) : undefined;
      if (!!nameSpace && !!formikRef?.current?.values && !!matchingParcel?.id) {
        const { resetForm, values } = formikRef.current;
        resetForm({ values: setIn(values, nameSpace, matchingParcel) });
        toast.dark('Found matching parcel within PIMS. Form data will be pre-populated.', {
          autoClose: 7000,
        });
      } else {
        const parcelDataLayerResponse = parcelsService.findByPid(parcelPid);
        handleParcelDataLayerResponse(parcelDataLayerResponse, dispatch);
      }
    });
  };

  const handleGeocoderChanges = async (data: IGeocoderResponse, nameSpace?: string) => {
    if (!!formikRef?.current && data) {
      const currentValues = !!nameSpace
        ? { ...getIn(formikRef.current.values, nameSpace) }
        : { ...formikRef.current.values };
      const newValues = {
        ...currentValues,
        latitude: data.latitude,
        longitude: data.longitude,
        address: {
          ...getIn(formikRef.current.values, nameSpace ?? '').address,
          line1: data.fullAddress,
        },
      };

      const administrativeArea = data.administrativeArea
        ? lookupCodes.find(code => {
            return (
              code.type === API.AMINISTRATIVE_AREA_CODE_SET_NAME &&
              code.name === data.administrativeArea
            );
          })
        : undefined;

      if (administrativeArea) {
        newValues.address.administrativeArea = administrativeArea.name;
      } else {
        newValues.address.administrativeArea = '';
      }

      const province = data.provinceCode
        ? lookupCodes.find(code => {
            return code.type === API.PROVINCE_CODE_SET_NAME && code.code === data.provinceCode;
          })
        : undefined;

      if (province) {
        newValues.address.provinceId = province.code;
        newValues.address.province = province.name;
      }

      // Ask geocoder for PIDs associated with this address
      let parcelPid: string = '';
      if (data.siteId) {
        try {
          const { pids } = await api.getSitePids(data.siteId);
          parcelPid = pids && pids.length > 0 ? pids[0] : '';
        } catch (error) {
          console.error('Failed to get pids');
        }
      }
      if (
        formikRef.current.values.pid &&
        parcelPid !== '' &&
        formikRef.current.values.pid !== parcelPid
      ) {
        setPidSelection({
          showPopup: true,
          geoPID: parcelPid.replace(/(\d{3})(\d{3})(\d{3})/, '$1-$2-$3'),
        });
      } else if (formikRef.current.values.pid && parcelPid === '') {
        newValues.pid = formikRef.current.values.pid;
      } else {
        if (parcelPid?.length) {
          newValues.pid = parcelPid;
          fetchPimsParcelByPid(newValues, parcelPid, nameSpace);
        } else {
          if (data.latitude && data.longitude) {
            const parcelDataLayerResponse = parcelsService.findOneWhereContains({
              lat: data.latitude,
              lng: data.longitude,
            } as LatLng);
            handleParcelDataLayerResponse(parcelDataLayerResponse, dispatch);
          }
        }
      }
    }
  };
  return { handleGeocoderChanges, pidSelection, setPidSelection };
};

export default useGeocoder;
