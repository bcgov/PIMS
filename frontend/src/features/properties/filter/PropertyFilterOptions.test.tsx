import { fireEvent, render, waitFor } from '@testing-library/react';
import { Formik } from 'formik';
import { PimsAPI, useApi } from 'hooks/useApi';
import React from 'react';

import { PropertyFilterOptions } from '.';
import { IPropertyFilter } from './IPropertyFilter';

jest.mock('axios');
jest.mock('@react-keycloak/web');
jest.mock('hooks/useApi');
(useApi as unknown as jest.Mock<Partial<PimsAPI>>).mockReturnValue({
  searchAddress: jest.fn(() => {
    return Promise.resolve([]);
  }),
});

const filter: IPropertyFilter = {
  searchBy: 'address',
  pid: '',
  address: '',
  administrativeArea: '',
  projectNumber: '',
  agencies: '',
  classificationId: '',
  minLotSize: '',
  maxLotSize: '',
  rentableArea: '',
};

const component = (filter: IPropertyFilter) => (
  <div>
    <Formik initialValues={filter} onSubmit={() => {}}>
      <PropertyFilterOptions />
    </Formik>
  </div>
);

describe('PropertyFilterOptions', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders for address', () => {
    const { container } = render(component(filter));
    expect(container.firstChild).toMatchSnapshot('address');
  });

  it('renders for pid', () => {
    const { container } = render(component({ ...filter, searchBy: 'pid' }));
    expect(container.firstChild).toMatchSnapshot('pid');
  });

  it('make request to geocoder', async () => {
    const value = '12345';
    const api = useApi();

    const { container } = render(component(filter));
    const address = container.querySelector('input[name="address"]');

    await waitFor(() => {
      fireEvent.change(address!, {
        target: {
          value: value,
        },
      });
      expect(api.searchAddress).toBeCalledTimes(1);
      expect(api.searchAddress).toBeCalledWith(value);
    });
  });

  it('less than 5 characters', async () => {
    const value = '1234';
    const api = useApi();

    const { container } = render(component(filter));
    const address = container.querySelector('input[name="address"]');

    await waitFor(() => {
      fireEvent.change(address!, {
        target: {
          value: value,
        },
      });
      expect(api.searchAddress).toBeCalledTimes(0);
    });
  });

  it('do not make request to geocoder', async () => {
    const value = '12345';
    const api = useApi();

    const { container } = render(component({ ...filter, address: value }));
    const address = container.querySelector('input[name="address"]');

    await waitFor(() => {
      fireEvent.change(address!, {
        target: {
          value: value,
        },
      });
      expect(api.searchAddress).toBeCalledTimes(0);
    });
  });
});
