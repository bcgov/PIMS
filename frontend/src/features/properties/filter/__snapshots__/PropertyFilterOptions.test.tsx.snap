// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyFilterOptions renders for address: address 1`] = `
.c0 .form.Control {
  width: 100%;
  height: 100%;
  border: none;
  font-size: 16px;
  box-sizing: border-box;
  flex: 1 1;
  min-width: 0;
  margin-bottom: 0;
}

.c0 .suggestionList {
  background-color: white;
  border-style: solid;
  border-color: lightslategray;
  border-width: thin;
  font-size: 14px;
  text-align: left;
  max-height: 600px;
  margin: 0;
  padding: 0;
  position: absolute;
  z-index: 999;
  top: 40px;
  left: -50px;
}

.c0 option {
  padding: 10px 5px;
  word-wrap: break-word;
  width: 400px;
  overflow: hidden;
  cursor: pointer;
  color: #494949;
}

.c0 option:hover {
  background-color: #007bff;
  color: #fff;
}

<div>
  <div
    class="input-group"
  >
    <div
      class="input-group"
    >
      <div
        class="select-row row"
        controlid="input-searchBy"
      >
        <div
          class="select-column col-md-auto"
        >
          <select
            class="form-select form-control"
            name="searchBy"
          >
            <option
              class="option"
              value="address"
            >
              Address
            </option>
            <option
              class="option"
              value="pid"
            >
              PID/PIN
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="input-group-content"
    >
      <div
        class="c0 GeocoderAutoComplete"
      >
        <div
          class=""
        >
          <div>
            <input
              class="form-control"
              field="address"
              id="input-address"
              name="address"
              placeholder="Enter an address"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PropertyFilterOptions renders for pid: pid 1`] = `
<div>
  <div
    class="input-group"
  >
    <div
      class="input-group"
    >
      <div
        class="select-row row"
        controlid="input-searchBy"
      >
        <div
          class="select-column col-md-auto"
        >
          <select
            class="form-select form-control"
            name="searchBy"
          >
            <option
              class="option"
              value="address"
            >
              Address
            </option>
            <option
              class="option"
              value="pid"
            >
              PID/PIN
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="input-group-content"
    >
      <div
        class="input-flex"
        id="input-pid"
      >
        <div>
          <input
            class="form-control"
            name="pid"
            placeholder="Enter a PID or PIN"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
